### PR TITLE
Coroutine improvements

### PR DESCRIPTION
**Pull request by rokups**
tl;dr;

* Using coroutines no longer requires `fasm` - platform assembler is used.
* Added masm assembler support. Thats what is used on windows to assemble platform code.
* Added `-d:nimCoroutinesPlatformJumps` to allow using platform `setjmp`/`longjmp`, otherwise bundled ones are used but they do not save floating point registers.
* Fixed bugs and crashes as well as cleaned up code.
* Added a workaround for glibc security feature which prevented use of custom allocated stacks.

# MASM support

I added masm support and removed automatic use of fasm. On windows when using `vcc` or `icl` platform code in `.asm` files will be used. On other platforms `.S` files will be used. That includes unixes and mingw. Code in `.S` files properly handles microsoft and sysv ABIs in case you are wondering. Now using coroutines no more depends on external tool.

# -d:nimCoroutinesPlatformJumps

It was mentioned in IRC that plan is to get rid of `msvcrt.dll` dependency. Assembly files ship with custom `setjmp` and `longjmp` implementations and they are used by default. However these implementations do not save/restore floating point registers. Defining `nimCoroutinesPlatformJumps` makes compiler use `setjmp` and `longjmp` from platform library in case extra features are wanted. Also fixed a bug on x64 where two registers were not properly saved.

# coroExecWithStack()

Replaced `coroSwitchStack()` and `coroRestoreStack()` with `coroExecWithStack()`. This fixes multiple issues where compiler would place unwanted code between these calls or after `coroRestoreStack()` call use registers that were set up before the call using wrong stack. `coroExecWithStack()` does not return as well. Coroutine is terminated with `suspendPrivate(0, 2)` which basically `jongjmp()`s to main loop, removes coroutine from the scheduler list and frees resources.

# Gremlins

~One thing i am unsure about is workaround for masm executable name. Problem is that older msvc compilers used to have different assembler exe names - `ml.exe` for 32bit and `ml64.exe` for 64bit. Some time after vs2010 this changed and now its always `ml.exe`. In order to support older compilers i added workaround [here](https://github.com/nim-lang/Nim/compare/devel...rokups:feature/coroutines?expand=1#diff-d7d89c7eb9bee7025e7b6bdc0bcd657bR598). In case you have better idea how to handle this please comment.~ Two assemblers were added instead of one. No nasty workarounds.

`./koch boot -d:release -d:nimCoroutines` still fails. I think there is bug somewhere in how gc handles multiple stacks. This one is for the future PRs.

Added `-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0` defines to disable some security features of glibc. In some cases glibc would initialize stack to some magic values which are checked on `longjmp()`. This only happens if `-d:nimCoroutinesPlatformJumps` is used.

# Testing

Test was successfully running `coro.nim` under these conditions:

Windows: all combinations of debug/release + opt:speed/size/none + msvc/mingw + 32bit/64bit
Linux: all combinations of debug/release + opt:speed/size/none + gcc/clang + 32bit/64bit
MacOS: did not test, no hardware/vm, someone should volunteer for this one.

It is safe to say that coroutines now officially **somewhat work**.
ARM when? I do not ARM. Maybe later ;)
If you want to try it - just `nim c -r lib/pure/coro.nim`.

### CODE REVIEW

**Comment by cheatfate** [comment]
MacOS tests can be done via Travis CI.

**Comment by rokups** [comment]
Also bundled `setjmp`/`longjmp` implementations do not save/restore floating point registers. Adding this feature is blocked by #5315. Saving/restoring fpu registers requires `JmpBuf` to be 16-byte aligned on some systems otherwise it crashes. MinGW64 using platform jumps triggers this crash currently.

**Comment by cheatfate** [comment]
Is there any reason for this attributes change?
`lib/pure/coro.nim 100644 → 100755`

**Comment by rokups** [comment]
Latest commit stores `xmm6` and `xmm7` on windows builds as described [here](https://en.wikipedia.org/wiki/X86_calling_conventions#Microsoft_x64_calling_convention). Wikipedia sometimes is not most accurate source so in case anyone thinks this is wrong please do tell. It Appears to be working ok. Stack alignment is taken care of in bundled `setjmp`/`longjmp` as well.

After tuning `cellsets.nim` i managed to achieve proper alignment but mingw64 build with `nimCoroutinesPlatformJumps` crashes anyway. Reason for this is `msvcrt.dll`:
```asm
_setjmp:
    mov     [rcx], rdx    ; stores rdx register at very start of the jmp_buf buffer. That's fine.
; ...
```
```asm
__longjmp_internal:
    xor     r10, r10
    cmp     [rcx], r10   ; If stored rdx was not 0 jump
    jnz     destination
; ...
destination:
; ...
    call    RtlUnwindEx_0
```
So in short - if `rdx` was not 0 when calling `msvcrt.dll:_setjmp` then it crashes. My conclusion is that it must be a bug of MinGW. Do not use `-d:nimCoroutinesPlatformJumps` with MinGW64.

# Tests

List of tests that were successful:

vc2010 and vc2017 x86
```bat
..\..\bin\nim.exe c -r --cc:vcc --cpu:amd64 -d:debug --opt:none coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:amd64 -d:debug --opt:speed coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:amd64 -d:debug --opt:size coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:amd64 -d:release --opt:none coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:amd64 -d:release --opt:speed coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:amd64 -d:release --opt:size coro.nim || GOTO :EOF

..\..\bin\nim.exe c -r --cc:vcc --cpu:i386 -d:debug --opt:none coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:i386 -d:debug --opt:speed coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:i386 -d:debug --opt:size coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:i386 -d:release --opt:none coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:i386 -d:release --opt:speed coro.nim || GOTO :EOF
..\..\bin\nim.exe c -r --cc:vcc --cpu:i386 -d:release --opt:size coro.nim || GOTO :EOF
```
vc2017 needs some work for `ml.exe` to be happy on x64.

MinGW:
```sh
test() {
    nim c -r "$@" coro.nim
    if [[ "$?" != "0" ]];
    then
        echo "Failed: $@"
        exit -1
    fi
    rm -rf lib
}

test --opt:none  --cpu:amd64 -d:debug
test --opt:speed --cpu:amd64 -d:debug
test --opt:size  --cpu:amd64 -d:debug
test --opt:none  --cpu:amd64 -d:release
test --opt:speed --cpu:amd64 -d:release
test --opt:size  --cpu:amd64 -d:release

test --opt:none  --cpu:i386 -d:debug
test --opt:speed --cpu:i386 -d:debug
test --opt:size  --cpu:i386 -d:debug
test --opt:none  --cpu:i386 -d:release
test --opt:speed --cpu:i386 -d:release
test --opt:size  --cpu:i386 -d:release
```

Linux:
```sh
test --cc:gcc   --opt:none  --cpu:amd64 --passC:-m64 --passL:-m64 -d:debug
test --cc:gcc   --opt:speed --cpu:amd64 --passC:-m64 --passL:-m64 -d:debug
test --cc:gcc   --opt:size  --cpu:amd64 --passC:-m64 --passL:-m64 -d:debug
test --cc:gcc   --opt:none  --cpu:amd64 --passC:-m64 --passL:-m64 -d:release
test --cc:gcc   --opt:speed --cpu:amd64 --passC:-m64 --passL:-m64 -d:release
test --cc:gcc   --opt:size  --cpu:amd64 --passC:-m64 --passL:-m64 -d:release

test --cc:gcc   --opt:none  --cpu:i386  --passC:-m32 --passL:-m32 -d:debug
test --cc:gcc   --opt:speed --cpu:i386  --passC:-m32 --passL:-m32 -d:debug
test --cc:gcc   --opt:size  --cpu:i386  --passC:-m32 --passL:-m32 -d:debug
test --cc:gcc   --opt:none  --cpu:i386  --passC:-m32 --passL:-m32 -d:release
test --cc:gcc   --opt:speed --cpu:i386  --passC:-m32 --passL:-m32 -d:release
test --cc:gcc   --opt:size  --cpu:i386  --passC:-m32 --passL:-m32 -d:release

test --cc:clang --opt:none  --cpu:amd64 --passC:-m64 --passL:-m64 -d:debug
test --cc:clang --opt:speed --cpu:amd64 --passC:-m64 --passL:-m64 -d:debug
test --cc:clang --opt:size  --cpu:amd64 --passC:-m64 --passL:-m64 -d:debug
test --cc:clang --opt:none  --cpu:amd64 --passC:-m64 --passL:-m64 -d:release
test --cc:clang --opt:speed --cpu:amd64 --passC:-m64 --passL:-m64 -d:release
test --cc:clang --opt:size  --cpu:amd64 --passC:-m64 --passL:-m64 -d:release

test --cc:clang --opt:none  --cpu:i386  --passC:-m32 --passL:-m32 -d:debug
test --cc:clang --opt:none  --cpu:i386  --passC:-m32 --passL:-m32 -d:release
```

Failing on Linux:
```sh
test --cc:clang --opt:speed --cpu:i386  --passC:-m32 --passL:-m32 -d:debug
test --cc:clang --opt:size  --cpu:i386  --passC:-m32 --passL:-m32 -d:debug
test --cc:clang --opt:speed --cpu:i386  --passC:-m32 --passL:-m32 -d:release
test --cc:clang --opt:size  --cpu:i386  --passC:-m32 --passL:-m32 -d:release
```
Crash happens here:
```asm
public add_A9bDQDKw5SEqD3OvtOnC1bg
add_A9bDQDKw5SEqD3OvtOnC1bg:
; ...
 [esp+70h], xmm1    ; esp=F748FE54
; ...
call    ___sprintf_chk
```
And C code:
```cpp
N_NIMCALL(void, add_A9bDQDKw5SEqD3OvtOnC1bg)(NimStringDesc** result0, NF x0) {
	TY_eVNFTutn6un5gcq48fQLdg buf0;
	NI n0;
	int LOC1;
	NIM_BOOL hasdot0;
	memset((void*)buf0, 0, sizeof(buf0));
	LOC1 = (int)0;
	LOC1 = sprintf(((NCSTRING) (buf0)), "%.16g", x0);
// ....
```
Maybe another compiler bug?


Hey @cheatfate  thank you for spotting that. I likely edited file from windows VM through samba share and thats what happens. Ill fix that when rebasing on to devel HEAD.

**Comment by rokups** [comment]
Rebased on devel, cleaned up that file mode, but most importantly turns out that `ml64.exe` workaround is not needed. Someone misinformed me that later visual studio versions renamed 64 bit assembler executable. That is not true. Thing is visual studio 2017 does not create shortcut for 64 bit developer command prompt so it is 32 bit by default. One is supposed to switch executing `"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64`. Also file search reveals `ml64.exe` being available. Also `vccexe.exe` gets in a way and sets vs2010 toolchain even if nim is called from vs2017 command prompt. Any way to control this?

**Comment by Araq** [comment]
> Maybe another compiler bug?

Is that still an issue? If so, what's wrong with the compiler?

**Comment by rokups** [comment]
Yes, its still a problem. I messed up problematic assembly when making previous comment. Problem is this:
```
public add_A9bDQDKw5SEqD3OvtOnC1bg
add_A9bDQDKw5SEqD3OvtOnC1bg:
; ...
movaps [esp+70h], xmm1    ; esp=F748FE54
; ...
call    ___sprintf_chk
```
clang generates code that tries to move `xmm1` to memory location that is not 16-byte aligned. It happens in nim compiler produced function `add_A9bDQDKw5SEqD3OvtOnC1bg` but i do not believe that code is faulty. Pasted relevant snipped of that proc in previous comment as well.

**Comment by rokups** [comment]
Last commits fixed remaining bugs known to me. If anyone manages to break coroutines now that would be helpful.

**Comment by rokups** [comment]
I got bit carried away and completely reworked coroutines once more. Its true i was kind of going in circles a bit but all that effort materialized into best coroutine version so far. Since it is a massive change i decided to make a new branch before overwriting stuff in this pull request.

Please check out https://github.com/rokups/Nim/tree/feature/coroutines2

Code highlights:
* ucontext support for switching contexts. I made this default because it does not require any additional platform-specific code. **Arm support** should be easy with this one.
* Fibers support on windows. I made this required on windows because it seems to do bit more behind the scenes than just switching register values. This is the safe approach.
* setjmp/longjmp support. Like before we can use libc or bundled versions on unixes.

Due to using fibers on windows we do not need any platform-specific code on this OS as well. Masm support is pointless by now but i still left it there. I guess it does not hurt to support assembler of popular compiler.

Any objections for stuff in that branch? Do we want that and i should overwrite this PR with `coroutines2` branch?

@Araq please review [this](https://github.com/rokups/Nim/commit/3838eaf415a396581ee466653376bf15c6ba0a3c#diff-99d47b8f45928bcbfd50e23d140ac969L269) change. I changed back gc'ing registers to your `c_setjmp` method. I thought about it and figured gc should pick those registers up because main thread stack is still added to `gch.stack` chain. Please verify if this assumption is correct.

**Comment by cheatfate** [comment]
First of all we need to test exceptions, iterators, gc memory leaks.

**Comment by rokups** [comment]
Since noone objected i pushed new work. Iterators and exceptions work in coroutines now, however gc support is still broken and i think it needs to be reworked from scratch. The way it is now, if merged it would be no worse and even better than what is in main nim repo at the moment. As for gc part rework - ill get to it next, but it might get slow.

**Comment by rokups** [comment]
@Araq pushed an attempt to realtime support for coroutines: [#3b3b04a](https://github.com/nim-lang/Nim/pull/5317/commits/3b3b04addc05bed7cbc0a288c8fe0c862c1e3eff). Think it is correct?

**Comment by Araq** [comment]
>  Think it is correct?

Yes, looks good.

**Comment by rokups** [comment]
What is this 3049.2 build on https://travis-ci.org/nim-lang/Nim/builds/202960164 ? It is marked as failing but 3049.1 is marked as successful. So is CI failing there or what? Here on github it says that all checks have passed.

File: compiler/commands.nim [hunk 1]
```diff
@@ -640,10 +640,6 @@ proc processSwitch(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
   of "experimental":
     expectNoArg(switch, arg, pass, info)
     gExperimentalMode = true
-  of "assembler":
```
Comment (Araq) [inline compiler/commands.nim]: this means it's a breaking change. Has 'assembler' ever been documented?

Comment (rokups) [inline compiler/commands.nim]: I never documented it so probably not.
